pkg/statsd/handler_backend.go

```diff
@@ -192,7 +192,10 @@ func (bh *BackendHandler) DispatchEvent(ctx context.Context, e *gostatsd.Event)
 			bh.eventWg.Add(eventsDispatched - len(bh.backends))
 			return
 		case bh.concurrentEvents <- struct{}{}:
-			go bh.dispatchEvent(ctx, backend, e)
+			// Creates a new context for dispatching the event.
+			// We create a new one otherwise it uses the request context which is cancelled as soon as this function returns.
+			timeoutCtx, _ := context.WithTimeout(context.Background(), 20*time.Second)
```

Comment:
We ditch the CancelFunc because if we defer it here, we still have the same issue that the event may not be sent. We want this func to return so that we don't block the caller.
From https://golang.org/pkg/context/#pkg-overview
>  Failing to call the CancelFunc leaks the child and its children until the parent is canceled or the timer fires.

Reply:
I think we should be explicit - we can use a pattern like:
```
go func() {
    bh.internalDispatchEvent(timeoutCtx, backend, e)
    cancel()
}()
```

... or, pass the cancel to `internalDispatchEvent` ...

```
func (bh *BackendHandler) internalDispatchEvent(ctx context.Context, backend gostatsd.Backend, e *gostatsd.Event, cancel func()) {
  defer cancel()
...
}
```

The former is a bit cleaner from "who is responsible", but the latter is neater, because the `go func() {}()` syntax is a bit wordy.  Either will do.

Reply:
```go
go func() {
    timeoutCtx, cancelTimeout := context.WithTimeout(context.Background(), 20*time.Second)
    defer cancelTimeout()
    bh.internalDispatchEvent(timeoutCtx, backend, e)
}()
```